client/executor/wasmtime/src/imports.rs

```diff
@@ -0,0 +1,327 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::state_holder::StateHolder;
+use sc_executor_common::error::WasmError;
+use sp_wasm_interface::{Function, Value, ValueType};
+use std::any::Any;
+use std::rc::Rc;
+use wasmtime::{
+	Callable, Extern, ExternType, Func, FuncType, ImportType, Limits, Memory, MemoryType, Module,
+	Trap, Val,
+};
+
+pub struct Imports {
+	/// Contains the index into `externs` where the memory import is stored if any. `None` if there
+	/// is none.
+	pub memory_import_index: Option<usize>,
+	pub externs: Vec<Extern>,
+}
+
+/// Goes over all imports of a module and prepares a vector of `Extern`s that can be used for
+/// instantiation of the module. Returns an error if there are imports that cannot be satisfied.
+pub fn resolve_imports(
+	state_holder: &StateHolder,
+	module: &Module,
+	host_functions: &[&'static dyn Function],
+	heap_pages: u32,
+	allow_missing_func_imports: bool,
+) -> Result<Imports, WasmError> {
+	let mut externs = vec![];
+	let mut memory_import_index = None;
+	for import_ty in module.imports() {
+		if import_ty.module() != "env" {
+			return Err(WasmError::Other(format!(
+				"host doesn't provide any imports from non-env module: {}:{}",
+				import_ty.module(),
+				import_ty.name()
+			)));
+		}
+
+		let resolved = match import_ty.name() {
+			"memory" => {
+				memory_import_index = Some(externs.len());
+				resolve_memory_import(module, import_ty, heap_pages)?
+			}
+			_ => resolve_func_import(
+				module,
+				state_holder,
+				import_ty,
+				host_functions,
+				allow_missing_func_imports,
+			)?,
+		};
+		externs.push(resolved);
+	}
+	Ok(Imports {
+		memory_import_index,
+		externs,
+	})
+}
+
+fn resolve_memory_import(
+	module: &Module,
+	import_ty: &ImportType,
+	heap_pages: u32,
+) -> Result<Extern, WasmError> {
+	let memory_ty = match import_ty.ty() {
+		ExternType::Memory(memory_ty) => memory_ty,
+		_ => {
+			return Err(WasmError::Other(format!(
+				"this import must be of memory type: {}:{}",
+				import_ty.module(),
+				import_ty.name()
+			)))
+		}
+	};
+
+	let at_least = memory_ty.limits().min().saturating_add(heap_pages);
+	if let Some(max) = memory_ty.limits().max() {
+		if at_least > max {
+			return Err(WasmError::Other(format!(
+				"incremented number of pages by heap_pages (total={}) is more than maximum requested\
+				by the runtime wasm module {}",
+				at_least,
+				max,
+			)));
+		}
+	}
+	let limits = Limits::at_least(at_least);
```

Comment:
```suggestion
	let limits = Limits::new(at_least, at_least);
```

We don't want to have the memory grow above the set heap pages?

Reply:
This is what wasmi does so yeah. Although I wonder if we should reconsider this!

Reply:
So when we use this for validating a parachain, we only want that the validation gets 1GB of memory and can not use more. I think that is something we should enforce/be able to enforce.